File: rust/src/location.rs
pub struct Location {
  pub line: usize,
  pub column: usize,
}

impl Location {
  pub fn new(line: usize, column: usize) -> Self {
    Self { line, column }
  }

  pub fn copy(&self) -> Self {
    Self { line: self.line, column: self.column }
  }

  pub fn to_string(&self) -> String {
    format!(
      "#<Location ({}:{})>",
      self.line,
      self.column,
    )
  }
}

File: node/lib/location.ts
export class Location {
  readonly line: number;
  readonly column: number;

  constructor(line: number, column: number) {
    this.line = line;
    this.column = column;
  }

  toHash(): Record<string, number> {
    return { line: this.line, column: this.column };
  }

  toJSON(): Record<string, number> {
    return this.toHash();
  }

  treeInspect(): string {
    return `(${this.line}:${this.column})`;
  }

  inspect(): string {
    return `#<Location ${this.treeInspect()}>`;
  }

  toString(): string {
    return this.inspect();
  }
}
